src/pybind/mgr/volumes/fs/misc_util.py

```diff
@@ -0,0 +1,215 @@
+        self.rfiles_h = 0   # num of regfiles handled
+        self.dirs_h = 0     # num of dirs handled
+        self.slinks_h = 0   # num of symlinks handled
+
+        self.op = op
+        # creating self.oped so that we can pring log messages accordingly.
+        self.oped_str = self._get_operated_str() # oped = operated
+
+    @property
+    def percent(self):
+        if self.size_t == 0:
+            return 100 if self.size_h == 0 else 0
+        return ((self.size_h/self.size_t) * 100)
+
+    @property
+    def progress_fraction(self):
+        '''
+        For progress MGR module (mgr/progress), value of attribute "progress"
+        should be in range of 0.0 to 1.0. 0.0 represents 0% progress and 1.0
+        represents 100% progress.
+        '''
+        return round(self.percent, 2)/100
+
+    def _get_operated_str(self):
+        '''
+        Return past tense form of the operation for which this instance is
+        being used.
+        '''
+        if self.op == 'copy':
+            return 'copied'
+        elif self.op == 'purge':
+            return 'purged'
+        else:
+            log.critical('Class Stats is being used for an operation it '
+                         'doesn\'t recognize')
+
+    def get_size_stat(self, human=True):
+        if human:
+            x, y = get_size_h(self.size_h), get_size_h(self.size_t)
+        else:
+            x, y = self.size_h, self.size_t
+
+        return f'{x}/{y}'
+
+    def get_rfiles_stat(self, human=True):
+        if human:
+            x, y = get_num_h(self.rfiles_h), get_num_h(self.rfiles_t)
+        else:
+            x, y = self.rfiles_h, self.rfiles_t
+
+        return f'{x}/{y}'
+
+    def get_dirs_stat(self, human=True):
+        if human:
+            x, y = get_num_h(self.dirs_h), get_num_h(self.dirs_t)
+        else:
+            x, y = self.dirs_h, self.dirs_t
+
+        return f'{x}/{y}'
+
+    def get_slinks_stat(self, human=True):
+        if human:
+            x, y = get_num_h(self.slinks_h), get_num_h(self.slinks_t)
+        else:
+            x, y = self.slinks_h, self.slinks_t
+
+        return f'{x}/{y}'
+
+    def get_progress_report(self, human=True):
+        return {
+            'percentage completed': self.percent,
+            f'amount {self.oped_str}': self.get_size_stat(human=human),
+            f'regfiles {self.oped_str}': self.get_rfiles_stat(human=human),
+            f'dirs {self.oped_str}': self.get_dirs_stat(human=human),
+            f'symlinks {self.oped_str}': self.get_slinks_stat(human=human)
+        }
+
+    def get_progress_report_str(self):
+        progress_report_str = ''
+        for k, v in self.get_progress_report().items():
+            progress_report_str += f'{k} = {v}\n'
+
+        # remove extra new line character at end, since print and log.xxx()
+        # methods automatically add a newline at the end.
+        return progress_report_str[:-1]
+
+    def log_handled_amount(self):
```

Comment:
Yes, in case cloning fails for some reason it would be easier to figure at what point did this issue occur.

Reply:
But, this does not log any failure, just the stats. How is this exactly helping with diagnosing a failure?

Reply:
When a failure occurs it becomes easier when we can see how much progress was made through logs. This info won't be available elsewhere. This also serves as a checkpoint in finding the point of failure when a crash occurs.

Reply:
> How does that help? It's more important that the reason for the failure is known (which already gets logged). Could you elaborate on how this information is useful (say this clone fails at ~10% completion at path /x/y/z)?

There were times during testing when clone operation was stuck because cloner threads were stuck or because CephFS was full. In such times, when no progress is made over a considerable stretch of time, it tells that issue is not with patches on this PR but elsewhere.

> Sorry, which crash?

I am not referring to particular kind of crash but rather all kinds of crash that I saw during testing.

Reply:
> > How does that help? It's more important that the reason for the failure is known (which already gets logged). Could you elaborate on how this information is useful (say this clone fails at ~10% completion at path /x/y/z)?
> 
> There were times during testing when clone operation was stuck because cloner threads were stuck

Stuck where? Did you run into some bug? Details would help here...

> or because CephFS was full. In such times, when no progress is made over a considerable stretch of time, it tells that issue is not with patches on this PR but elsewhere.

This isn't a compelling reason to log stats on each iteration. Also, if this point is considered w.r.t. comment https://github.com/ceph/ceph/pull/54620#discussion_r1444164281 (which is still under discussion), I think explicit tracking can be done away wait.

> 
> > Sorry, which crash?
> 
> I am not referring to particular kind of crash but rather all kinds of crash that I saw during testing.

MDS crash, client crash, python code?